### PR TITLE
Add PcapNG Support to PacketFu

### PR DESCRIPTION
**Pull request by claudijd**
This is a work in progress branch/PR for adding pcap-ng support to packetfu.

Here's a list of things we'll need to do:

1. Create some pcap-ng files for testing and comparison [DONE]
2. Write the logic for parsing/writing pcap-ng
3. Write unit-tests to make sure it works ok
4. Get people to try it out, give us feedback with pcap-ng as an option, but not the default
5. Make pcap-ng the default in development
6. Release a new major version of PacketFu that would support pcap-ng as the default

### CODE REVIEW

**Comment by sdaubert** [comment]
@claudijd need some help for 2. ? Seems to be interesting to do.

**Comment by claudijd** [comment]
@sdaubert yeah, if you're interested that would be great.  This has been a feature that's been sitting around for a while and it would be a good to make some progress on it.  Have a look at the specification [here](http://xml2rfc.tools.ietf.org/cgi-bin/xml2rfc.cgi?url=https://raw.githubusercontent.com/pcapng/pcapng/master/draft-tuexen-opsawg-pcapng.xml&modeAsFormat=txt/ascii&type=ascii).

Where I'm at currently is just grok'ing the specification and thinking about how I want to approach a prototype.  If you have suggestions you are welcome to add them here.  Also, if you write any code feel free to send as a pull request to this branch or link as a gist.

Thanks!

**Comment by sdaubert** [comment]
@claudijd I would define a file object which holds all section objects, which, in turn holds all others object as Interface Description, Enhanced Packet and Simple Packet.

For reading, I would ignore all other blocks and also all options but ts_resolv from IDB (if we need to keep timestamps).
Do PacketFu need to support writing pcap-ng file?

**Comment by claudijd** [comment]
Feature TODO Update:

- Complete the logic for parsing pcap-ng files:
  - add support for Simple Packet Blocks.
  - add support for timestamps.
  - add support for writing Pcap-NG files.
- Write the logic for writing pcap-ng files
- Make sure we have solid unit-test coverage for PcapNG to make sure it works
- Get people to try it out, give us feedback with pcap-ng as an option, but not the default
- Make pcap-ng the default in development
- Release a new major version of PacketFu that would support pcap-ng as the default

**Comment by claudijd** [comment]
@sdaubert regarding your question about writing PCAPNG files, I think it would be the definition of having PCAPNG full-support that we can both read/write PCAPNG.  From what I've seen from other implementations, they tend to only support PCAPNG reading because writing is likely going to require some invasive refactoring of logic to retain the necessary information, like interface data.  Reading seems like the first most logical place to start as you've already made great headway with, and it's very possible the TODO items I've listed above may include engaging people early to try out the reader implementation (as an option, but not a default) before we tackle the write portion of things.

**Comment by claudijd** [comment]
@sdaubert a nice array of PCAP-NG test files we can use to torture test the reader implementation when we're ready => https://github.com/hadrielk/pcapng-test-generator

**Comment by sdaubert** [comment]
@claudijd We should be ready for that.
But before achieving that, i would prefer complete Pcapng::File API to mimic PcapFile one.

**Comment by sdaubert** [comment]
@claudijd we should provide a mean to get raw unknown blocks when reading a Pcap-NG file. The easiest way seems to add new instance variable array in Pcapng::File to store these blocks. This array may be named `@unknown_blocks`. An attribute reader should also be provided.
This will help us to do torture tests.

**Comment by claudijd** [comment]
@sdaubert I like your suggestion and agree the instance variable will be a nice way to have a soft-failure, but allow us to monitor what we are unable to parse.

**Comment by sdaubert** [comment]
@claudijd in fact, as all blocks depend to SHB in logical hierarchy, i think unknown blocks should be recorded in SHB object instead of File one.
About maintaining offsets, i don't known. UnknownBlock already maintain raw content of unsupported blocks. So we may already get their content. Is this what you mean by "_find them later when we want to do testing_" ?

**Comment by claudijd** [comment]
@sdaubert yeah, but you're right, we could easily find the offset for troubleshooting purposes like you said.  We should be fine.

**Comment by claudijd** [comment]
**TODO** before landing in master...

- [ ] @sdaubert overall review
- [ ] @claudijd overall review
- [ ] @todb overall review
- [ ] Add a script in ./examples that we can give beta testers that want to try out the PCAP-NG feature

File: lib/packetfu/pcapng/file.rb [hunk 1]
```diff
@@ -0,0 +1,316 @@
+require_relative 'shb'
+
+module PacketFu
+  module PcapNG
+
+    # PcapNG::File is a comple Pcap-NG file handler.
```
Comment (claudijd) [inline lib/packetfu/pcapng/file.rb]: Minor typo "comple" => "complete"

File: lib/packetfu/pcapng/file.rb [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+      # the Pcapng::File object.
+      def read(str)
+        PacketFu.force_binary(str)
+        io = StringIO.new(str)
+        parse_section(io)
+        self
+      end
+
+      # Clear the contents of the Pcapng::File prior to reading in a new string.
+      # This string should contain a Section Header Block and a Interface Description
```
Comment (claudijd) [inline lib/packetfu/pcapng/file.rb]: Grammar "and a Interface" => "and an Interface"

**Comment by claudijd** [comment]
@sdaubert I think we're just about ready to land this feature in master.  If you can give the PR one last review to make sure you're happy and we land this.  Once I get your "ok" then we'll merge it.

As a follow up to this, I'm planning on a pre/dev/beta gem release of PacketFu to allow people to start playing with the PCAP-NG functionality and provide feedback.

**Comment by sdaubert** [comment]
@claudijd that's ok for me.

**Pull request merged**